test/gpu/native/studies/sort/radixSortGpu.chpl

```diff
@@ -0,0 +1,279 @@
+          var prefixSumBuffer = prefixSum;
+          foreach idx in 0..<buckets/(2*offset) {
+            const i = idx*2*offset;
+            const t = prefixSumBuffer[i+offset-1];
+            prefixSum[i+offset-1] = prefixSumBuffer[i+2*offset-1];
+            prefixSum[i+2*offset-1] = prefixSum[i+2*offset-1] + t;
+          }
+          offset = offset >> 1;
+        }
+        hostprefixSum = prefixSum;
+      }
+  }
+
+  proc arrScan(ref count: [?D] uint, ref inputArr: [] uint, exp: int, bitMask: int, buckets: int){
+    // Count the number of occurrences of each bitsAtATime-bit chunk in the array
+    // at the current exp position
+    on here.gpus[1]{
+      var gpuCount : [D] uint; // Initialized all 0s
+      var gpuInputArr = inputArr;
+      foreach i in gpuInputArr {
+        assertOnGpu();
+        gpuAtomicAdd(gpuCount[((i>>exp) & bitMask):int], 1);
+        // count[((i>>exp) & bitMask):int] += 1;
+      }
+      count = gpuCount;
+    }
+    if scanTest then
+      writeln(" Count: ", count);
+
+
+    if scanType == "hillisSteele" then
+      hillisSteeleScan(count, buckets);
+    else if scanType == "blelloch" then
+      blellochScan(count, buckets);
+    else if scanType == "serial" then
+      serialScan(count, buckets);
+
+    if scanTest then
+      writeln("Pf sum: ",count);
+  }
+
+  proc radixSort(inputArr: [] uint){
+    if (noisy){
+        writeln("Bits at a time: ", bitsAtATime);
+        writeln("Buckets: ", buckets);
+        writeln("Bit mask: ", bitMask);
+        writeln("Input: ", inputArr);
+    }
+
+    // Create the output array and copy input
+    var outputArr: [0..<inputArr.size] uint;
+
+    // Create the count and prefix sum arrays once
+    // This was we can reuse it for each iteration of radix Sort
+    var prefixSum: [0..<buckets] uint;
+
+    // Ceiling on number of iterations based on max element in array
+    // Can reduce do this faster?
+    var max = inputArr[0];
+    for i in 1..<inputArr.size {
+      if inputArr[i] > max {
+        max = inputArr[i];
+      }
+    }
+
+    // Number of iterations is the number of bits in the max element divided by number of bits we sort at a time
+    var exp = 0;
+    while (max> 0){
+      if(noisy) then
+        writeln("       Exp: ", exp);
+
+      arrScan(prefixSum, inputArr, exp, bitMask, buckets);
+
+      if noisy then
+        writeln("       Prefix sum: ", prefixSum);
+      // Use the prefix sum to reorder the array at the exp position
+      // This is done by iterating through the array backwards
+      // and placing each element in the correct position in the output array
+
+      /*
+      on here.gpus[1]{
+        var gpuPrefixSum = prefixSum;
+        var gpuInputArr = inputArr;
+        var gpuOutputArr : outputArr.type;
+        foreach i in inputArr{
+          assertOnGpu();
+          const idx = ((i>>exp) & bitMask):int;
+          const arrIdx = gpuAtomicFetchAndAdd(gpuPrefixSum[idx], 1); // Need this operation
```

Comment:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicadd, it certainly does fetch and return. We just need to wire it correctly in our library. It'd be great if you can take up on that.

You can merge the PR as-is in the interim, and add that later on.

Reply:
Andy got me started with the atomics. I'm working on adding that functionality right now to all atomic operations, not just add, for both nvidia and amd. 

I'll merge this PR and then add a patch to use the atomics once I'm done with that task.

Reply:
Well I ended up actually merging the atomics PR before this one 😅 So I guess I'll update this now and retest